jumanji/environments/routing/macvrp/env.py

```diff
@@ -0,0 +1,600 @@
+
+        # Main vehicle spec
+        main_vehicle_dict = {
+            "positions": vehicle_position,
+            "local_times": vehicle_local_time,
+            "capacities": vehicle_capacity,
+        }
+        main_vehicle_spec = specs.Spec(
+            ObsVehicle, "MainVehicleSpec", **main_vehicle_dict
+        )
+
+        return specs.Spec(
+            Observation,
+            "ObservationSpec",
+            nodes=nodes_spec,
+            windows=windows_spec,
+            coeffs=penality_spec,
+            other_vehicles=other_vehicle_spec,
+            main_vehicles=main_vehicle_spec,
+            action_mask=action_mask,
+        )
+
+    def action_spec(self) -> specs.BoundedArray:
+        """
+        Returns the action spec.
+
+        Returns:
+            action_spec: a `specs.BoundedArray` spec.
+        """
+
+        return specs.BoundedArray(
+            (self._num_vehicles,),
+            jax.numpy.int16,
+            0,
+            self._num_customers + 1,
+            "actions",
+        )
+
+    def render(self, state: State) -> Optional[NDArray]:
+        """Render the given state of the environment. This rendering shows the layout of the tour so
+            far with the cities as circles, and the depot as a square.
+
+        Args:
+            state: environment state to render.
+
+        Returns:
+            rgb_array: the RGB image of the state as an array.
+        """
+        return self._viewer.render(state)
+
+    def animate(
+        self,
+        states: Sequence[State],
+        interval: int = 200,
+        save_path: Optional[str] = None,
+    ) -> matplotlib.animation.FuncAnimation:
+        """Create an animation from a sequence of states.
+
+        Args:
+            states: sequence of `State` corresponding to subsequent timesteps.
+            interval: delay between frames in milliseconds, default to 200.
+            save_path: the path where the animation file should be saved. If it is None, the plot
+                will not be saved.
+
+        Returns:
+            animation that can export to gif, mp4, or render with HTML.
+        """
+        return self._viewer.animate(states, interval, save_path)
+
+    def _update_state(self, state: State, actions: chex.Array) -> State:
+        """
+        Updates the state of the environment.
+
+        Args:
+            state: State object containing the dynamics of the environment.
+            next_node: int, index of the next node to visit.
+
+        Returns:
+            state: State object corresponding to the new state of the environment.
+        """
+
+        # Convert actions to int16.
+        next_nodes = jax.numpy.int16(actions)
+
+        # Zero any node selections if the node has zero demand or does not have enough
+        # capacity left.
+        next_nodes = (
+            next_nodes
+            * (state.vehicles.capacities > state.nodes.demands[next_nodes])
+            * (state.nodes.demands[next_nodes] > 0)
```

Comment:
This is the same as action mask, right ? In some other envs, the action_mask is stored in the state so that it can be used here to mask invalid actions. This avoids recomputing the mask here and duplication of code, so easier to change the logic later if needed. What do you think ?

Reply:
Updated.